pkg/rancher/migrations.go

```diff
@@ -132,6 +141,33 @@ func createOrUpdateConfigMap(configMapClient controllerv1.ConfigMapClient, cm *v
 	return err
 }
 
+func versionTombstone(configMapController controllerv1.ConfigMapController) error {
+	cm, err := getConfigMap(configMapController, rancherVersionTombstoneConfig)
+	if err != nil || cm == nil {
+		return err
+	}
+
+	if !semver.IsValid(rancherversion.Version) {
+		logrus.Errorf("rancher version %s is not semver compliant, skipping tombstone validation", rancherversion.Version)
+		return nil
+	}
+
+	lastVersion := cm.Data[rancherVersionKey]
+	if lastVersion != "" {
+		if !semver.IsValid(lastVersion) {
+			logrus.Errorf("Previous Rancher version %s is not semver compliant, skipping tombstone validation", lastVersion)
+			lastVersion = ""
```

Comment:
It looks like it should just return as the error message indicates.  Otherwise, it is possible to compare an empty string with the value of `rancherversion.Version` at line 161


```suggestion
			return nil
```

Reply:
We still need to
```
cm.Data[rancherVersionKey] = rancherversion.Version
	return createOrUpdateConfigMap(configMapController, cm)
```
on 167-168 because otherwise we would perform this check everytime on startup, which is not necessary. Generally speaking, my idea is that once you set the server version to something that is not valid semver, it basically clears the last one since there is no real way to know what it was. This is somewhat to safeguard the case where a user manually modifies the config map, clearing the value because semver comparison is impossible. Here is an example:
`v2.9.0 -> vtestversion -> v2.9.1`
After upgrading to `vtestversion`, the normal flow would be rancher gets the key from the config map as `v2.9.0` compares it to `vtestversion`, and does not update the config map, instead logging an error and returning `nil`. However, were a user to manually set the `rancherversion` key in the configmap to `vtestversion`, if this function were to return `nil` the version would never update. In other words, under your suggestion, if an invalid semver version was ever set in the configmap, Rancher would never clear it. Now, a proper solution would be to implement something in the webhook, but to the best of my knowledge the webhook may not exist because I have created a new category of preflight migrations to support this which runs before the `systemcharts` controller exists which deploys the webhook, and I don't believe we have prior art for the capability to watch specific objects in the webhook.

Reply:
Thanks for the explanation. It now makes sense to me.  Could you add a clarifying comment to the code?

Reply:
The purpose of this PR is to prevent possible data loss scenarios which may occur when Rancher is downgraded without using BRO. This is probably a shortsighted approach, but generally speaking the most disruptive behavior is going to occur when the `OnLeader` function is called. From a provisioning perspective, the level of operational failure can range from an unnecessary reprovisioning cycle, cascading provisioning errors, incorrect configs being deployed, to in some extreme cases clusters being entirely wiped out.

Also to clarify on:
> I actually would agree with @jiaqiluo 's original comment, and I'm a bit confused as to what this is preventing. From what I can see, we need to check the configmap on every start of the leader pod with or without this change. I can understand us wanting to override this to a valid version if it's currently an invalid version, but I don't think that's what this is doing.

That's not the point I'm making. My point is if we `return nil` here, then we never update the value in the configmap, regardless of what the current version is. The logic is as follows:
- Get the current rancher version
  - if it's not valid, skip validation & return
- Get the last known version
  - if it's not valid, skip validation
- Update the value in the configmap to the current rancher version

Adding in `return nil` if the last known version is not valid means we never update the last known version, meaning it's always invalid, which seems wrong.